core/lib/contract_verifier/src/resolver/mod.rs

```diff
@@ -138,116 +164,119 @@ pub(crate) trait Compiler<In>: Send + fmt::Debug {
     ) -> Result<CompilationArtifacts, ContractVerifierError>;
 }
 
-/// Default [`CompilerResolver`] using pre-downloaded compilers in the `/etc` subdirectories (relative to the workspace).
 #[derive(Debug)]
-pub(crate) struct EnvCompilerResolver {
-    home_dir: PathBuf,
+pub struct ResolverMultiplexer {
+    resolvers: Vec<Arc<dyn CompilerResolver>>,
 }
 
-impl Default for EnvCompilerResolver {
-    fn default() -> Self {
+impl ResolverMultiplexer {
+    pub fn new(resolver: Arc<dyn CompilerResolver>) -> Self {
         Self {
-            home_dir: Workspace::locate().core(),
+            resolvers: vec![resolver],
         }
     }
-}
 
-impl EnvCompilerResolver {
-    async fn read_dir(&self, dir: &str) -> anyhow::Result<Vec<String>> {
-        let mut dir_entries = fs::read_dir(self.home_dir.join(dir))
-            .await
-            .context("failed reading dir")?;
-        let mut versions = vec![];
-        while let Some(entry) = dir_entries.next_entry().await? {
-            let Ok(file_type) = entry.file_type().await else {
-                continue;
-            };
-            if file_type.is_dir() {
-                if let Ok(name) = entry.file_name().into_string() {
-                    versions.push(name);
-                }
-            }
-        }
-        Ok(versions)
+    pub fn with_resolver(mut self, resolver: Arc<dyn CompilerResolver>) -> Self {
+        self.resolvers.push(resolver);
+        self
     }
 }
 
 #[async_trait]
-impl CompilerResolver for EnvCompilerResolver {
+impl CompilerResolver for ResolverMultiplexer {
     async fn supported_versions(&self) -> anyhow::Result<SupportedCompilerVersions> {
-        Ok(SupportedCompilerVersions {
-            solc: self
-                .read_dir("etc/solc-bin")
-                .await
-                .context("failed reading solc dir")?,
-            zksolc: self
-                .read_dir("etc/zksolc-bin")
-                .await
-                .context("failed reading zksolc dir")?,
-            vyper: self
-                .read_dir("etc/vyper-bin")
-                .await
-                .context("failed reading vyper dir")?,
-            zkvyper: self
-                .read_dir("etc/zkvyper-bin")
-                .await
-                .context("failed reading zkvyper dir")?,
-        })
+        let mut versions = SupportedCompilerVersions::default();
+        for resolver in &self.resolvers {
+            versions.merge(resolver.supported_versions().await?);
+        }
+        Ok(versions)
     }
 
+    /// Resolves a `solc` compiler.
     async fn resolve_solc(
         &self,
         version: &str,
     ) -> Result<Box<dyn Compiler<SolcInput>>, ContractVerifierError> {
-        let solc_path = CompilerType::Solc.bin_path(&self.home_dir, version).await?;
-        Ok(Box::new(Solc::new(solc_path)))
+        let mut last_error = Err(ContractVerifierError::UnknownCompilerVersion(
```

Comment:
The problem is that some methods take version as `&str`, and some as `&ZkCompilerVersions`. I've initially tried to write a macro but decided that it's somewhat ugly, and it's not that much of boilerplate.

I agree that it can be refactored, but my few attempts have failed (e.g. it's not easy to make this function generic, and making everything accept `ZkCompilerVersions` makes little sense).

Do you insist on fixing this?

Reply:
No, it's a small concern.